samza-test/src/main/java/org/apache/samza/test/framework/system/InMemoryOutputDescriptor.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.test.framework.system;
+
+import com.google.common.base.Preconditions;
+import org.apache.samza.operators.descriptors.base.stream.OutputDescriptor;
+import org.apache.samza.operators.descriptors.base.system.SystemDescriptor;
+import org.apache.samza.serializers.NoOpSerde;
+
+/**
+ * A descriptor for an in memory output stream.
+ * <p>
+ * An instance of this descriptor may be obtained from an appropriately configured {@link InMemorySystemDescriptor}.
+ * <p>
+ * Stream properties configured using a descriptor override corresponding properties provided in configuration.
+ *
+ * @param <StreamMessageType> type of messages in this stream.
+ */
+public class InMemoryOutputDescriptor<StreamMessageType>
+    extends OutputDescriptor<StreamMessageType, InMemoryOutputDescriptor<StreamMessageType>> {
+
+  private int partitionCount;
+
+  /**
+   * Constructs an {@link OutputDescriptor} instance.
+   * @param streamId id of the stream
+   * @param systemDescriptor system descriptor this stream descriptor was obtained from
+   */
+  InMemoryOutputDescriptor(String streamId, SystemDescriptor systemDescriptor) {
+    super(streamId, new NoOpSerde<>(), systemDescriptor);
+  }
+
+  /**
+   * Creates an in memory output stream with given partition count
+   * @param partitionCount partition count of output stream
+   * @return this output descriptor
+   */
+  public InMemoryOutputDescriptor<StreamMessageType> withPartitionCount(int partitionCount) {
```

Comment:
There seems to be an inconsistency between this and other streams that Samza works with. For other output streams, Samza assumes that the stream already exists with some partition count, so partition count is not needed in the descriptor.
Maybe it would be more consistent if TestRunner.addOutputStream accepted the partition count as an argument for creating the stream.

Reply:
PartitionCount is an attribute of a Stream, so should be described along with InMemoryStreams (in the descriptor). 

Streams are defined by systems and not Samza. Kafka streams may have different attributes as compared to Brooklyn or InMemoryStreams. 

Hence this implementation feels correct to me.